File: ignite-extensions/modules/ignite-graph/works-ignite-common/src/main/scala/de/kp/works/ignite/IgniteUtil.scala
package de.kp.works.ignite

import de.kp.works.ignite.graph.ElementType
import org.apache.ignite.binary.BinaryObject
import org.apache.ignite.cache.{CacheMode, QueryEntity}
import org.apache.ignite.configuration.CacheConfiguration
import org.apache.ignite.{Ignite, IgniteCache}

import java.util

object IgniteUtil {

  def createCacheIfNotExists(ignite: Ignite, table: String, cfg: CacheConfiguration[String, BinaryObject]): Unit = {
    val exists: Boolean = ignite.cacheNames.contains(table)
    if (!exists) ignite.createCache(cfg)
  }

  def getOrCreateCache(ignite: Ignite, table: String, namespace: String): IgniteCache[String, BinaryObject] = {

    val exists: Boolean = ignite.cacheNames.contains(table)
    if (exists) {
      ignite.cache(table)
    }
    else {
      createCache(ignite, table, namespace)
    }
  }

  def createCache(ignite: Ignite, table: String, namespace: String): IgniteCache[String, BinaryObject] = {
    createCache(ignite, table, namespace, CacheMode.REPLICATED)
  }

  def createCache(ignite: Ignite, table: String, namespace: String, cacheMode: CacheMode): IgniteCache[String, BinaryObject] = {
    /*
     * Derive the table type from the provided
     * table name
     */
    val tableType = if (table == namespace + "_" + IgniteConstants.EDGES)
      ElementType.EDGE

    else if (table == namespace + "_" + IgniteConstants.VERTICES)
      ElementType.VERTEX

    else

      throw new Exception("Table '" + table + "' is not supported.")

    val cfg = createCacheCfg(table, tableType, cacheMode)
    ignite.createCache(cfg)

  }

  def createCacheCfg(table: String, tableType: ElementType, cacheMode: CacheMode): CacheConfiguration[String, BinaryObject] = {
    /*
     * Defining query entities is the Apache Ignite
     * mechanism to dynamically define a queryable
     * 'class'
     */
    val qe = buildQueryEntity(table, tableType)
    val qes = new util.ArrayList[QueryEntity]
    qes.add(qe)
    /*
     * Specify Apache Ignite cache configuration; it is
     * important to leverage 'BinaryObject' as well as
     * 'setStoreKeepBinary'
     */
    val cfg = new CacheConfiguration[String, BinaryObject]
    cfg.setName(table)

    cfg.setStoreKeepBinary(false)
    cfg.setIndexedTypes(classOf[String], classOf[BinaryObject])

    cfg.setCacheMode(cacheMode)
    cfg.setQueryEntities(qes)

    cfg
  }

  /**
   * This method supports the creation of different
   * query entities (or cache schemas)
   */
  def buildQueryEntity(table: String, elementType: ElementType): QueryEntity = {

    val qe = new QueryEntity
    /*
     * The key type of the Apache Ignite cache is set to
     * [String], i.e. an independent identity management is
     * used here
     */
    qe.setKeyType("java.lang.String")
    /*
     * The 'table' is used as table name in select statement
     * as well as the name of 'ValueType'
     */
    qe.setValueType(table)
    /*
     * Define fields for the Apache Ignite cache that is
     * used as one of the data backends.
     */
    if (elementType == ElementType.EDGE)
      qe.setFields(buildEdgeFields)

    else if (elementType == ElementType.VERTEX)
      qe.setFields(buildVertexFields())

    else
      throw new Exception("Table '" + table + "' is not supported.")

    qe
  }

  def buildEdgeFields: util.LinkedHashMap[String, String] = {

    val fields = new util.LinkedHashMap[String, String]
    /*
     * The edge identifier used by TinkerPop to
     * identify an equivalent of a data row
     */
    fields.put(IgniteConstants.ID_COL_NAME, "java.lang.String")
    /*
     * The edge identifier type to reconstruct the
     * respective value. IgniteGraph supports [Long]
     * as well as [String] as identifier.
     */
    fields.put(IgniteConstants.ID_TYPE_COL_NAME, "java.lang.String")
    /*
     * The edge label used by TinkerPop and IgniteGraph
     */
    fields.put(IgniteConstants.LABEL_COL_NAME, "java.lang.String")
    /*
     * The `TO` vertex description
     */
    fields.put(IgniteConstants.TO_COL_NAME, "java.lang.String")
    fields.put(IgniteConstants.TO_TYPE_COL_NAME, "java.lang.String")
    /*
     * The `FROM` vertex description
     */
    fields.put(IgniteConstants.FROM_COL_NAME, "java.lang.String")
    fields.put(IgniteConstants.FROM_TYPE_COL_NAME, "java.lang.String")
    /*
             * The timestamp this cache entry has been created.
             */
    fields.put(IgniteConstants.CREATED_AT_COL_NAME, "java.lang.Long")
    /*
     * The timestamp this cache entry has been updated.
     */
    fields.put(IgniteConstants.UPDATED_AT_COL_NAME, "java.lang.Long")
    /*
     * The property section of this cache entry
     */
    fields.put(IgniteConstants.PROPERTY_KEY_COL_NAME, "java.lang.String")
    fields.put(IgniteConstants.PROPERTY_TYPE_COL_NAME, "java.lang.String")
    /*
     * The serialized property value
     */
    fields.put(IgniteConstants.PROPERTY_VALUE_COL_NAME, "java.lang.String")
    fields
  }

  def buildVertexFields(): util.LinkedHashMap[String, String] = {

    val fields: util.LinkedHashMap[String, String] = new util.LinkedHashMap[String, String]()
    /*
     * The vertex identifier used by TinkerPop to identify
     * an equivalent of a data row
     */
    fields.put(IgniteConstants.ID_COL_NAME, "java.lang.String")
    /*
     * The vertex identifier type to reconstruct the
     * respective value. IgniteGraph supports [Long]
     * as well as [String] as identifier.
     */
    fields.put(IgniteConstants.ID_TYPE_COL_NAME, "java.lang.String")
    /*
     * The vertex label used by TinkerPop and IgniteGraph
     */
    fields.put(IgniteConstants.LABEL_COL_NAME, "java.lang.String")
    /*
     * The timestamp this cache entry has been created.
     */
    fields.put(IgniteConstants.CREATED_AT_COL_NAME, "java.lang.Long")
    /*
     * The timestamp this cache entry has been updated.
     */
    fields.put(IgniteConstants.UPDATED_AT_COL_NAME, "java.lang.Long")
    /*
     * The property section of this cache entry
     */
    fields.put(IgniteConstants.PROPERTY_KEY_COL_NAME, "java.lang.String")
    fields.put(IgniteConstants.PROPERTY_TYPE_COL_NAME, "java.lang.String")
    /*
     * The serialized property value
     */
    fields.put(IgniteConstants.PROPERTY_VALUE_COL_NAME, "java.lang.String")
    fields
  }

}

File: ignite-extensions/modules/ignite-graph/works-ignite-common/src/main/java/de/kp/works/ignite/IgniteUtil.java
package de.kp.works.ignite;


import org.apache.ignite.configuration.*;
import org.apache.ignite.internal.processors.cache.GridCacheDefaultAffinityKeyMapper;
import org.apache.ignite.binary.*;
import org.apache.ignite.*;

import de.kp.works.ignite.graph.*;
import org.apache.ignite.cache.*;


import java.util.*;

public final class IgniteUtil{
	public static IgniteUtil MODULE$ = new IgniteUtil();
	/**
	 *  只使用key的前面冒号开头的字符串前缀
	 * @author admin
	 *
	 */
	public static class KeyPrefixAffinityKeyMapper extends GridCacheDefaultAffinityKeyMapper{		
		private static final long serialVersionUID = 1L;

		@Override public Object affinityKey(Object key) {
			 if(key instanceof String) {
				 String strKey = key.toString();
				 int pos = strKey.indexOf(':');
				 if(pos>0) {
					 return strKey.substring(0,pos);
				 }
				 return key;
			 }
			 return super.affinityKey(key);
		 }
				
	}
    
    public void createCacheIfNotExists(final Ignite ignite, final String table, final CacheConfiguration<String, BinaryObject> cfg) {
        final boolean exists = ignite.cacheNames().contains(table);
        if (!exists) {
            ignite.createCache(cfg);
        }
    }
    
    public IgniteCache<String, BinaryObject> getOrCreateCache(final Ignite ignite, final String table, final String namespace) {
        final boolean exists = ignite.cacheNames().contains(table);
        return (exists ? ignite.cache(table) : this.createCache(ignite, table, namespace));
    }
    
    public IgniteCache<String, BinaryObject> createCache(final Ignite ignite, final String table, final String namespace) {
        return this.createCache(ignite, table, namespace, CacheMode.PARTITIONED);
    }
    
    public IgniteCache<String, BinaryObject> createCache(final Ignite ignite, final String name, final String namespace, final CacheMode cacheMode) {
        
        ElementType elementType = null;
        /*
         * Retrieve element type from provided
         * cache (table) name
         */
        if (name.equals(namespace + "_" + IgniteConstants.EDGES)) {
            elementType = ElementType.EDGE;
        }
        else if (name.equals(namespace + "_" + IgniteConstants.VERTICES)) {
            elementType = ElementType.VERTEX;
        }
        else if (name.startsWith(namespace + "_")) {
            elementType = ElementType.DOCUMENT;
        }
        else if (name.startsWith(namespace + ".")) {
            elementType = ElementType.DOCUMENT;
        }
        else {
            elementType = ElementType.UNDEFINED;
            throw new RuntimeException("table name not support!"+name);
        }
        final ElementType tableType = elementType;
        final CacheConfiguration<String, BinaryObject> cfg = this.createCacheCfg(name, tableType, cacheMode);
        return ignite.createCache(cfg);
    }
    
    public CacheConfiguration<String, BinaryObject> createCacheCfg(final String table, final ElementType tableType, final CacheMode cacheMode) {
        final QueryEntity qe = this.buildQueryEntity(table, tableType);
        final ArrayList<QueryEntity> qes = new ArrayList<>();
        qes.add(qe);
        final CacheConfiguration<String, BinaryObject> cfg = new CacheConfiguration<>();
        cfg.setName(table);
        cfg.setStoreKeepBinary(false);
        cfg.setIndexedTypes(new Class[] { String.class, BinaryObject.class });
        cfg.setCacheMode(cacheMode);
        cfg.setQueryEntities(qes);
        // add@byron
        cfg.setAffinityMapper(new KeyPrefixAffinityKeyMapper());
        // end@
        return cfg;
    }
    
    public QueryEntity buildQueryEntity(final String table, final ElementType elementType) {
        final QueryEntity qe = new QueryEntity();
        qe.setKeyType("java.lang.String");
        qe.setValueType(table);
        
        if (elementType == ElementType.EDGE) {
        	
        	qe.setFields(this.buildEdgeFields());
            qe.setIndexes(this.buildEdgeIndexs());
            return qe;
        }
        else if(elementType == ElementType.VERTEX) {
        	
        	qe.setFields(this.buildVertexFields());
            qe.setIndexes(this.buildVertexIndexs());
            return qe;
        } 
        if (elementType == null || !elementType.equals(ElementType.DOCUMENT)) {
            throw new RuntimeException(new StringBuilder().append("Table '").append(table).append("' is not supported.").toString());
        }
        qe.setFields(this.buildDocumentFields(qe));
        
        if(!table.endsWith("_index")) {
        	qe.setIndexes(this.buildDocumentIndexs());
        }
        return qe;
    }
    
    public LinkedHashMap<String, String> buildEdgeFields() {
        final LinkedHashMap<String, String> fields = new LinkedHashMap<>();
        fields.put("id", "java.lang.String");
        fields.put("id_type", "java.lang.String");
        fields.put("label", "java.lang.String");
        fields.put("to", "java.lang.String");
        fields.put("to_type", "java.lang.String");
        fields.put("source", "java.lang.String");
        fields.put("source_type", "java.lang.String");
        fields.put("created_at", "java.lang.Long");
        fields.put("updated_at", "java.lang.Long");
        fields.put("property_key", "java.lang.String");
        fields.put("property_type", "java.lang.String");
        fields.put("property_value", "java.lang.String");
        return fields;
    }
    
    public List<QueryIndex> buildEdgeIndexs() {
        final List<QueryIndex> indexes = new ArrayList<>();
        indexes.add(new QueryIndex("source"));
        indexes.add(new QueryIndex("to"));
        return indexes;
    }
    
    public LinkedHashMap<String, String> buildVertexFields() {
        final LinkedHashMap<String, String> fields = new LinkedHashMap<>();
        fields.put("id", "java.lang.String");
        fields.put("id_type", "java.lang.String");
        fields.put("label", "java.lang.String");
        fields.put("created_at", "java.lang.Long");
        fields.put("updated_at", "java.lang.Long");
        fields.put("property_key", "java.lang.String");
        fields.put("property_type", "java.lang.String");
        fields.put("property_value", "java.lang.String");
        return fields;
    }
    
    public List<QueryIndex> buildVertexIndexs() {
        final List<QueryIndex> indexes = new ArrayList<>();
        indexes.add(new QueryIndex("id"));
        indexes.add(new QueryIndex(Arrays.asList("label","property_key"),QueryIndexType.SORTED));        
        return indexes;
    }
    
    public LinkedHashMap<String, String> buildDocumentFields(QueryEntity qe) {
        final LinkedHashMap<String, String> fields = new LinkedHashMap<>();
        fields.put("id", "java.lang.String");
        fields.put("type", "java.lang.String");        
        fields.put("created_at", "java.lang.Long");
        fields.put("updated_at", "java.lang.Long");
        fields.put("name", "java.lang.String");
        fields.put("title", "java.lang.String");
        fields.put("author", "java.lang.String");
        
        qe.setKeyFieldName("id");
        return fields;
    }
    
    public List<QueryIndex> buildDocumentIndexs() {
        final List<QueryIndex> indexes = new ArrayList<>();
        indexes.add(new QueryIndex("id"));
        indexes.add(new QueryIndex("updated_at"));        
        return indexes;
    }
}

